crates/cli-support/src/js/mod.rs

```diff
@@ -1168,19 +1168,77 @@ impl<'a> Context<'a> {
         } else {
             ""
         };
-        self.global(&format!(
+
+        // The first implementation we have for this is to use
+        // `TextEncoder#encode` which has been around for quite some time.
+        let use_encode = format!(
             "
-            function passStringToWasm(arg) {{
                 {}
                 const buf = cachedTextEncoder.encode(arg);
                 const ptr = wasm.__wbindgen_malloc(buf.length);
                 getUint8Memory().set(buf, ptr);
                 WASM_VECTOR_LEN = buf.length;
                 return ptr;
-            }}
             ",
             debug
-        ));
+        );
+
+        // Another possibility is to use `TextEncoder#encodeInto` which is much
+        // newer and isn't implemented everywhere yet. It's more efficient,
+        // however, becaues it allows us to elide an intermediate allocation.
+        let use_encode_into = format!(
+            "
+                {}
+                let size = arg.length;
+                let ptr = wasm.__wbindgen_malloc(size);
+                let writeOffset = 0;
+                while (true) {{
+                    const view = getUint8Memory().subarray(ptr + writeOffset, ptr + size);
+                    const {{ read, written }} = cachedTextEncoder.encodeInto(arg, view);
+                    arg = arg.substring(read);
+                    writeOffset += written;
+                    if (arg.length === 0) {{
+                        break;
+                    }}
+                    ptr = wasm.__wbindgen_realloc(ptr, size, size * 2);
+                    size *= 2;
```

Comment:
If the allocator isn't a power-of-two allocator, this can overallocate. It is known that the max number of bytes that `encodeInto` can require is JavaScript string length times 3. (It follows that if the Wasm program knows the string to be short-lived on the Wasm side, it makes sense to just allocate for the ma length case up front and not reallocate.)

Reply:
Hm good point! Do you think it'd be worthwhile do do something like `malloc(string.length)`, and if that fails `realloc(string.length * 3)` followed by `realloc(the_actual_length)` and that'd be more appropriate than this loop?

Reply:
If the allocations can be assumed the be short-lived, I suggest going with `malloc(string.length * 3)` without a loop.

If they are assumed to be long-lived, I suggest coming up with a magic number `magic` that characterizes how much smaller the initial allocation needs to than the worst case for it to be worthwhile not to allocate for the worst case right away. 

Then:

```
let min = roundUpToBucketSize(string.length);
let max = string.length * 3;
if (min + magic > max) {
  // min is close enough to max anyway
  let m = malloc(max);
  let (read, written) = encodeInto(...);
  assert(read == string.length);
} else {
  let m = malloc(min);
  let (read, written) = encodeInto(...);
  if (read < string.length) {
     let inputLeft = string.length - read;
     let outputNeeded = inputLeft * 3;
     let m = realloc(m, written + outputNeeded);
     // ... encodeInto with input starting from read and output starting from written
  }
}
```

Reply:
Ok thanks for the info! I've opened up https://github.com/rustwasm/wasm-bindgen/issues/1313 to further track this.